src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java

```diff
@@ -216,13 +307,53 @@ private PaymentHandleResponse executeWithExceptionallyHandling(String paymentId,
-                            return PaymentHandleResponse.ofHttpStatusAndErrorMessage(HttpStatus.valueOf(restException.getResponsecode()),
-                                    format("%s=[%s] can't be processed, see the logs", paymentIdType, paymentId));
+                            return saveResponseToInterfaceInteraction(paymentId, paymentIdType, restException)
+                                    .toCompletableFuture().join();
                         }
                     }
-                    return PaymentHandleResponse.of400BadRequest(
-                            format("%s [%s] can't be processed, see the logs", paymentIdType, paymentId));
-                }).toCompletableFuture().join();
+                    CompletionStage<PaymentHandleResponse> paymentHandleResponseStage = CompletableFuture.completedFuture(PaymentHandleResponse.of400BadRequest(
+                            format("%s [%s] can't be processed, details: [%s]", paymentIdType, paymentId, throwable.getMessage())));
+                    return paymentHandleResponseStage.toCompletableFuture().join();
+                });
+        return exceptionally.toCompletableFuture().join();
+    }
+
+    private CompletionStage<PaymentHandleResponse> saveResponseToInterfaceInteraction(@Nullable String paymentId,
+                                                                                      @Nullable String paymentIdType,
+                                                                                      @Nonnull PayPalRESTException restException) {
+        CompletionStage<String> ctpPaymentIdStage = CompletableFuture.completedFuture(paymentId);
+        if (PAYPAL_PLUS_PAYMENT_ID.equals(paymentIdType)) {
+            // if it's paypal plus payment id and not ctp payment id,
+            // then fetch ctp payment to get ctp payment id
+            ctpPaymentIdStage = this.ctpFacade.getPaymentService()
+                    .getByPaymentMethodAndInterfaceId(PAYPAL_PLUS, paymentId)
+                    .thenApply(ctpPayment -> ctpPayment.map(Resource::getId).orElse(null));
+        }
 
+        CompletionStage<PaymentHandleResponse> paymentHandleResponseStage = ctpPaymentIdStage
+                .thenCompose(ctpPaymentId -> {
+                    AddInterfaceInteraction action = createAddInterfaceInteractionAction(restException.getDetails(), RESPONSE);
+                    return ctpFacade.getPaymentService().updatePayment(ctpPaymentId, Collections.singletonList(action))
+                            .thenApply(ignore -> {
+                                return PaymentHandleResponse.ofHttpStatusAndErrorMessage(HttpStatus.valueOf(restException.getResponsecode()),
+                                        format("%s=[%s] can't be processed, details: [%s]", paymentIdType, paymentId, restException.getMessage()));
+                            });
+                });
+        // todo: don't join here, but rather return completion stage
```

Comment:
looks like the _todo_ is outdated already

Reply:
https://github.com/commercetools/commercetools-paypal-plus-integration/pull/49/commits/b2fe4857f7f3e0aab5e02391040c519fda5043e3